Track 3 - Login mutation (#42)
indigotech/taq-github-bot: src/data/local/github/tracks-node/track-3/steps/step-2.md

# Step 1 - Integrating with the database

### Estimated time: 4 hours

> Now you're going to fully integrate the Login mutation with the database. Since we don't have a mutation to create users (yet), we should create one directly on the database, so we have the conditions to test it properly.

Reviewer: ```suggestion
Now you're going to fully integrate the Login mutation with the database. Since we don't have a mutation to create users (yet), we should create one directly on the database, so we have some conditions to test it properly.
```

> **NOTE:** One of the required data to create a user is the password. You should have read or heard already that storing users password in plain text on database is a bad ideia. If not, you're reading now: it **is** a bad ideia 🤦‍. The minimum security we should have on this case is to use a hash algorithm to store the password. This is not 100% secure method, but it helps already. Interested in this security matter? Check [this post](https://itnext.io/how-not-to-store-passwords-4955569e6e84) for additional details. Since this is only an onboard server, you don't need to have a ultra-master-blaster security method. Our main goal here is to show you the degrees of secutrity, and prevent you from beginning your projects with an ultra-master-blaster insecure method.

Reviewer: ```suggestion
**NOTE:** One of the required data to create a user is the password. You should have read or heard already that storing users password as plain text in the database is a bad idea. If not, you're reading now: it **is** a bad idea 🤦‍. The minimum security we should have on this case is to use a hash algorithm to store the password. This is not a 100% secure method, but it helps already. Interested in this security matter? Check [this post](https://itnext.io/how-not-to-store-passwords-4955569e6e84) for additional details. Since this is only an onboard server, you don't need to have a ultra-master-blaster security method. Our main goal here is to show you some levels of security and prevent you from beginning your projects with an ultra-master-blaster (we like that word) insecure method.
```

After creating a user, you should implement your mutation as follows:

1. Get the input e-mail received on the mutation and try to find one on database.

> 1. Check if the given password is correct. Remember that you should apply all the layers of security to the input password in oder to compare them correctly.

Reviewer: Acho que dá pra remover essa parte:

> Remember that you should apply all the layers of security to the input password in oder to compare them correctly

 porque conflita com o que está na nota ali em cima: 

> Since this is only an onboard server, you don't need to have a ultra-master-blaster security method.

Author: Essa parte foi pra lembrar de comparar as senhas com hash. Como ficou meio aberto q forma como cada um vai guardar a senha, eu escrevi pra aplicar "todos". Vou mudar pra algo que deixe isso mais claro sem conflitar.

> 1. Return the user info and a token, so user can use this token as an `Authorization` header for future requests. You can use a [JWT token](https://jwt.io/introduction/) for that, with a secret to encode it and a shot expiration, like 15 minutes.

Reviewer: ```suggestion
1. Return the user info and a token. About the token, all you have to worry about is that it must be a [JWT token](https://jwt.io/introduction/) with a secret to encode it and with an expiration time set to 15 minutes. Ask to a Taqtiler if you need any help here ;)
```

> On each of these steps, some error may occur. It's important in these cases to provide a good description to the client of went wrong. For example, if the password is wrong, you could return an error message like "Invalid credentials, please check your e-mail and password". 

Reviewer 2: "... of what went wrong"
